### PR TITLE
Switch from Unit of work to Services.

### PR DESCRIPTION
**Pull request by Sputnavi**
I have doubts about my implementation of services.

### CODE REVIEW

File: BicycleCompany.Models/Response/PartProblemForReadModel.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace BicycleCompany.Models.Response
+{
+    public class PartProblemForReadModel
+    {
+        public Guid Id { get; set; }
+        public PartForReadModel Part { get; set; }
```
Comment (AnnaGolosova) [inline BicycleCompany.Models/Response/PartProblemForReadModel.cs:8]: Let it be `PartDetailsModel`

File: BicycleCompany.BLL/appsettings.json [hunk 2]
```diff
@@ -8,6 +8,13 @@
...
   "ConnectionStrings": {
-    "sqlConnection": "server=(localdb)\\mssqllocaldb;database=BicycleCompany;Integrated Security=true"
+    "sqlConnection": "server=(localdb)\\mssqllocaldb;database=BicycleCompany;Integrated Security=true",
+    "dockerSqlConnection": "Server=db;Database=BicycleCompany;User=sa;Password=My_password_123;"
+  },
+  "JwtSettings": {
+    "validIssuer": "BicycleComapny",
+    "validAudience": "https://localhost:5001",
+    "expires": 10,
+    "key": "bicycle!company!secret!key!"
```
Comment (AnnaGolosova) [inline BicycleCompany.BLL/appsettings.json:18]: Move all secret values from project setting to environment settings. In the future, we will use some secret storage for this, like Azure KeyVault, to GitHub secrets.

File: BicycleCompany.Tests/ProblemsControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+            var fakeProblemService = new Mock<IProblemService>();
+            fakeProblemService.Setup(r => r.GetProblemListAsync(null, null))
+                .Returns(Task.FromResult(new List<ProblemForReadModel>()));
+            var controller = new ProblemsController(null, fakeProblemService.Object);
+
+            // Act
+            var result = await controller.GetProblemList(null);
+
+            // Assert
+            Assert.IsType<OkObjectResult>(result);
```
Comment (AnnaGolosova) [inline BicycleCompany.Tests/ProblemsControllerTests.cs:86]: You should also check here that you get an empty list.

Comment (AnnaGolosova) [reply BicycleCompany.Tests/ProblemsControllerTests.cs:86]: And check that result is not null

File: BicycleCompany.Tests/ProblemsControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+            var controller = new ProblemsController(null, fakeProblemService.Object);
+
+            // Act
+            var response = await controller.GetProblemList(null);
+
+            // Assert
+            Assert.IsType<OkObjectResult>(response);
+            var result = response as OkObjectResult;
+
+            Assert.IsType<List<ProblemForReadModel>>(result.Value);
```
Comment (AnnaGolosova) [inline BicycleCompany.Tests/ProblemsControllerTests.cs:105]: I can suggest using FluentAssertions for all asserts https://fluentassertions.com/  
At least you can try to use it for some tests, no need to replace old implementation for all tests

File: BicycleCompany.Tests/ProblemsControllerTests.cs [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+            var problems = result.Value as List<ProblemForReadModel>;
+
+            Assert.Equal(problems.Count, GetFakeProblemList().Count);
```
Comment (AnnaGolosova) [inline BicycleCompany.Tests/ProblemsControllerTests.cs:108]: You need to check here ALL fields of ALL items. To ensure that controller returns values in the correct format, and doesn't modify it.

File: BicycleCompany.Tests/ProblemsControllerTests.cs [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+            var controller = new ProblemsController(null, fakeProblemService.Object);
+
+            // Act
+            var response = await controller.GetProblem(guid);
+
+            // Assert
+            Assert.IsType<OkObjectResult>(response);
+
+            var result = response as OkObjectResult;
+            Assert.IsType<ProblemForReadModel>(result.Value);
```
Comment (AnnaGolosova) [inline BicycleCompany.Tests/ProblemsControllerTests.cs:131]: Same comment. Always check all data that you can

File: BicycleCompany.Tests/ProblemsControllerTests.cs [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        public async Task CreateProblem_ShouldThrowArgumentException_WithInvalidObject()
+        {
+            // Arrange
+            var controller = new ProblemsController(null, null);
+            controller.ModelState.AddModelError("Fields", "Required");
+
+            var problem = new ProblemForCreateModel();
+
+            // Act & Assert
+            await Assert.ThrowsAsync<ArgumentException>(async () =>
```
Comment (AnnaGolosova) [inline BicycleCompany.Tests/ProblemsControllerTests.cs:168]: You should check all error messages here

File: BicycleCompany.Tests/ProblemsControllerTests.cs [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+
+            var part = new PartDetailsForCreateModel();
+
+            // Act & Assert
+            await Assert.ThrowsAsync<ArgumentException>(async () =>
+                await controller.CreatePartForProblem(Guid.NewGuid(), part));
+        }
+
+        [Fact]
+        public async Task DeletePartForProblem_ShouldReturnNoContent_WithValidId()
```
Comment (AnnaGolosova) [inline BicycleCompany.Tests/ProblemsControllerTests.cs:297]: Please, order all your tests. and better, move all Delete... tests to a separate region. It makes me crazy to jump through all this file to find more tests about Delete... method

File: BicycleCompany.Tests/ProblemsControllerTests.cs [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+
+            var problem = new ProblemForCreateModel();
+
+            // Act & Assert
+            await Assert.ThrowsAsync<ArgumentException>(async () => 
+                await controller.CreateProblem(problem));
+        }
+
+        [Fact]
+        public async Task DeleteProblem_ShouldReturnNoContent_WithValidId()
```
Comment (AnnaGolosova) [inline BicycleCompany.Tests/ProblemsControllerTests.cs:173]: What's the difference between this test and `DeletePartForProblem_ShouldReturnNoContent_WithValidId` test?

**Review by AnnaGolosova — CHANGES_REQUESTED** [review]

File: BicycleCompany.Tests/ProblemsControllerTests.cs (hunk 9 above)
Comment (Sputnavi) [reply BicycleCompany.Tests/ProblemsControllerTests.cs:173]: The difference is that these tests for different methods of controller. `DeleteProblem` is used to delete a Problem and `DeletePartForProblem` deletes a part from Problem note.

File: BicycleCompany.Tests/ProblemsControllerTests.cs (hunk 8 above)
Comment (Sputnavi) [reply BicycleCompany.Tests/ProblemsControllerTests.cs:297]: Main answer below. I order my tests by Method Name. I think it's ok :)
So should I use `#Region` to separate tests?

**Pull request closed without merging**